File: reddit-swarm/src/pipeline.js
const { getDb } = require('./db');
const { createBatchFromSchedule, getNextTask, markTaskProcessing, markTaskDone, markTaskFailed } = require('./agents/orchestrator');
const { getOrCreatePersona, incrementPersonaPostCount } = require('./agents/persona-builder');
const { writePost } = require('./agents/content-writer');
const { reviewPost } = require('./agents/quality-reviewer');
const { config, logger } = require('../config');

/**
 * Run a single task through the full pipeline:
 * 1. Pick next pending task
 * 2. Assign a persona
 * 3. Write the post
 * 4. Review for quality
 * 5. Update metrics
 */
async function runOne() {
  // 1. Get next task
  const task = await getNextTask();
  if (!task) {
    logger.debug('No pending tasks in queue');
    return null;
  }

  logger.info(`=== Pipeline: Processing task ${task.task_id} ===`);
  await markTaskProcessing(task.task_id);

  try {
    // 2. Assign persona
    const persona = await getOrCreatePersona(task.params);

    // Update task with persona
    const db = await getDb();
    await db.collection('post_queue').updateOne(
      { task_id: task.task_id },
      { $set: { persona_id: persona.persona_id } }
    );

    // 3. Write post
    const post = await writePost(task, persona);

    // 4. Review post
    const review = await reviewPost(post);

    // 5. Update persona post count
    await incrementPersonaPostCount(persona.persona_id);

    // 6. Mark task done
    await markTaskDone(task.task_id);

    logger.info(
      `=== Pipeline: Task ${task.task_id} complete — post ${review.approved ? 'APPROVED' : 'REJECTED'} (score: ${review.score}) ===`
    );

    return { task, persona, post, review };
  } catch (err) {
    await markTaskFailed(task.task_id, err);
    logger.error(`Pipeline failed for task ${task.task_id}: ${err.message}`);
    return { task, error: err };
  }
}

/**
 * Run a batch of tasks sequentially.
 */
async function runBatch(count) {
  const total = count || config.pipeline.postsPerDay;
  logger.info(`Starting batch of ${total} tasks`);

  const results = {
    total: 0,
    approved: 0,
    rejected: 0,
    failed: 0,
    posts: [],
  };

  for (let i = 0; i < total; i++) {
    logger.info(`--- Batch progress: ${i + 1}/${total} ---`);
    const result = await runOne();

    if (!result) {
      logger.info('No more tasks to process — batch complete');
      break;
    }

    results.total++;
    if (result.error) {
      results.failed++;
    } else if (result.review?.approved) {
      results.approved++;
      results.posts.push(result);
    } else {
      results.rejected++;
      results.posts.push(result);
    }

    // Small delay between tasks to avoid rate limiting
    if (i < total - 1) {
      await new Promise((resolve) => setTimeout(resolve, 500));
    }
  }

  // Update daily metrics
  await updateMetrics(results);

  logger.info(
    `Batch complete: ${results.total} processed, ${results.approved} approved, ${results.rejected} rejected, ${results.failed} failed`
  );

  return results;
}

/**
 * Run daily batch: create tasks then process them.
 */
async function runDaily(count) {
  const total = count || config.pipeline.postsPerDay;
  logger.info(`=== Starting daily run for ${total} posts ===`);

  // Step 1: Create tasks
  await createBatchFromSchedule(total);

  // Step 2: Process all tasks
  const results = await runBatch(total);

  logger.info(`=== Daily run complete ===`);
  return results;
}

/**
 * Update the metrics collection with today's results.
 */
async function updateMetrics(batchResults) {
  const db = await getDb();
  const today = new Date();
  today.setHours(0, 0, 0, 0);

  const generated = batchResults.total || 0;
  const approved = batchResults.approved || 0;
  const rejected = (batchResults.rejected || 0) + (batchResults.failed || 0);
  const approvalRate = generated > 0 ? approved / generated : 0;

  // Calculate average quality score from posts
  let totalScore = 0;
  let scoreCount = 0;
  for (const r of batchResults.posts || []) {
    if (r.review?.score != null) {
      totalScore += r.review.score;
      scoreCount++;
    }
  }
  const avgScore = scoreCount > 0 ? totalScore / scoreCount : 0;

  await db.collection('metrics').updateOne(
    { date: today },
    {
      $set: {
        generated,
        approved,
        rejected,
        approval_rate: Math.round(approvalRate * 1000) / 1000,
        avg_quality_score: Math.round(avgScore * 1000) / 1000,
      },
      $setOnInsert: {
        date: today,
      },
    },
    { upsert: true }
  );

  logger.info(`Metrics updated: ${approved}/${generated} approved (${(approvalRate * 100).toFixed(1)}%), avg score: ${avgScore.toFixed(3)}`);
}

/**
 * Get current metrics summary.
 */
async function getMetricsSummary() {
  const db = await getDb();

  // Today's metrics
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const todayMetrics = await db.collection('metrics').findOne({ date: today });

  // Last 7 days
  const weekAgo = new Date(today);
  weekAgo.setDate(weekAgo.getDate() - 7);
  const weekMetrics = await db
    .collection('metrics')
    .find({ date: { $gte: weekAgo } })
    .sort({ date: -1 })
    .toArray();

  const weekTotals = weekMetrics.reduce(
    (acc, m) => ({
      generated: acc.generated + (m.generated || 0),
      approved: acc.approved + (m.approved || 0),
      rejected: acc.rejected + (m.rejected || 0),
    }),
    { generated: 0, approved: 0, rejected: 0 }
  );

  return {
    today: todayMetrics || { generated: 0, approved: 0, rejected: 0, approval_rate: 0, avg_quality_score: 0 },
    week: weekTotals,
    dailyBreakdown: weekMetrics,
  };
}

module.exports = {
  runOne,
  runBatch,
  runDaily,
  getMetricsSummary,
};

File: reddit-swarm/src/pipeline.py
"""
Iron Kinetic Reddit Swarm - Pipeline Runner
High-level pipeline for generating, reviewing, and storing posts.
"""

from datetime import datetime, timezone

from pymongo import MongoClient

from src.config import Config, log
from src.db import get_client, get_collection, close_connection
from src.swarm_agent import IronKineticRedditAgent
from src.swarm_coordinator import SwarmCoordinator
from src.quality_reviewer import QualityReviewer


def run_one() -> dict:
    """Pick next pending job, run full pipeline.

    Returns:
        Result dict with status and post info.
    """
    client = get_client()
    content_coll = get_collection(Config.COLLECTION_CONTENT)

    # Find oldest draft job
    job = content_coll.find_one(
        {"stato": "draft", "titolo": ""},
        sort=[("created_at", 1)],
    )

    if job is None:
        log.info("No pending jobs found.")
        return {"status": "no_jobs", "message": "No pending jobs"}

    agent_instance_id = job.get("agentInstanceId")
    context_data = job.get("datiContestoUsati", {})
    user_data = context_data.get("userData", {})
    subreddit = context_data.get("subreddit", "")

    log.info(
        "Running job %s: agent=%s subreddit=%s",
        job.get("jobId"), agent_instance_id, subreddit,
    )

    # Generate post via agent
    agent = IronKineticRedditAgent(agent_instance_id, client)
    agent.load_configuration()

    context = {
        "utente_tipo": user_data.get("utente_tipo", "intermediate"),
        "lingua": user_data.get("lingua", "en"),
        "subreddit": subreddit,
        "angolo_preferito": user_data.get("angolo_preferito"),
        **user_data,
    }

    result = agent.generate_post(context)

    if result.get("status") == "error":
        # Update job as failed
        content_coll.update_one(
            {"_id": job["_id"]},
            {"$set": {
                "stato": "fallito",
                "feedback.error": result.get("metadata", {}).get("error", "Unknown"),
            }},
        )
        return {"status": "error", "job_id": job.get("jobId")}

    titolo = result.get("titolo", "")
    corpo = result.get("corpo", "")

    # Quality review
    reviewer = QualityReviewer()
    review_result = reviewer.review({"titolo": titolo, "corpo": corpo})

    score = review_result.get("score", 0.0)
    approved = review_result.get("approved", False)
    notes = review_result.get("notes", [])

    # Update content document
    new_stato = "pubblicato" if approved else "draft"
    content_coll.update_one(
        {"_id": job["_id"]},
        {"$set": {
            "titolo": titolo,
            "corpo": corpo,
            "stato": new_stato,
            "qualitaScore": score,
            "feedback.review": review_result,
            "datiContestoUsati.privacy": result.get("metadata", {}).get("privacy", {}),
            "timestampPubblicazione": datetime.now(timezone.utc) if approved else None,
        }},
    )

    log.info(
        "Job %s complete: score=%.2f approved=%s",
        job.get("jobId"), score, approved,
    )

    return {
        "status": "approved" if approved else "review_needed",
        "job_id": job.get("jobId"),
        "score": score,
        "notes": notes,
        "titolo": titolo,
    }


def run_batch(count: int = 5) -> list:
    """Run N posts sequentially.

    Args:
        count: Number of posts to process.

    Returns:
        List of result dicts.
    """
    results = []
    log.info("Starting batch of %d posts...", count)

    for i in range(count):
        log.info("--- Batch post %d/%d ---", i + 1, count)
        result = run_one()
        results.append(result)

        if result.get("status") == "no_jobs":
            log.info("No more jobs available, stopping batch.")
            break

    approved_count = sum(1 for r in results if r.get("status") == "approved")
    log.info(
        "Batch complete: %d/%d approved",
        approved_count, len(results),
    )

    return results


def run_daily() -> dict:
    """Create and process daily batch.

    Creates POSTS_PER_DAY jobs across features and languages,
    then processes them through the pipeline.

    Returns:
        Summary dict.
    """
    client = get_client()
    coordinator = SwarmCoordinator(client)

    posts_per_day = Config.POSTS_PER_DAY
    log.info("Starting daily pipeline: %d posts", posts_per_day)

    # Define distribution of posts
    distribution = [
        {"feature": "referral", "lingua": "en", "weight": 0.2},
        {"feature": "referral", "lingua": "it", "weight": 0.15},
        {"feature": "predictive_curve", "lingua": "en", "weight": 0.15},
        {"feature": "predictive_curve", "lingua": "it", "weight": 0.1},
        {"feature": "meal_plan", "lingua": "en", "weight": 0.1},
        {"feature": "meal_plan", "lingua": "it", "weight": 0.08},
        {"feature": "check_in", "lingua": "en", "weight": 0.07},
        {"feature": "check_in", "lingua": "it", "weight": 0.05},
        {"feature": "subscription", "lingua": "en", "weight": 0.05},
        {"feature": "subscription", "lingua": "it", "weight": 0.03},
        {"feature": "generic_progress", "lingua": "en", "weight": 0.01},
        {"feature": "generic_progress", "lingua": "it", "weight": 0.01},
    ]

    subreddits_en = ["r/loseit", "r/progresspics", "r/fitness", "r/nutrition", "r/BodyweightFitness"]
    subreddits_it = ["r/italy", "r/progresspics", "r/fitness"]

    import random

    jobs_created = 0
    for _ in range(posts_per_day):
        # Weighted random selection
        r = random.random()
        cumulative = 0.0
        selected = distribution[0]
        for d in distribution:
            cumulative += d["weight"]
            if r <= cumulative:
                selected = d
                break

        feature = selected["feature"]
        lingua = selected["lingua"]
        subs = subreddits_en if lingua == "en" else subreddits_it
        subreddit = random.choice(subs)

        user_data = {
            "utente_tipo": random.choice(["beginner", "intermediate", "experienced"]),
            "lingua": lingua,
        }

        try:
            coordinator.request_post_generation(
                feature=feature,
                lingua=lingua,
                subreddit=subreddit,
                user_data=user_data,
                force_diversity=True,
            )
            jobs_created += 1
        except Exception as e:
            log.warning("Failed to create job: %s", e)

    log.info("Created %d jobs. Processing...", jobs_created)

    # Process all created jobs
    results = run_batch(jobs_created)

    approved = sum(1 for r in results if r.get("status") == "approved")

    return {
        "jobs_created": jobs_created,
        "jobs_processed": len(results),
        "approved": approved,
        "approval_rate": round(approved / max(1, len(results)), 3),
    }
